### DISCUSSION TITLE
Control the reticle with a pi as server/client

### DISCUSSION THREAD

**Discussion by lamko648** [Stellarium/stellarium#4001] [Q&A]
The program makes a tcp/ip connection and can send some data. But nothing happens in Stellarium.
I'm using C and the logfile from stellarium says :
TelescopeTCP( "New Telescope 1" )::prepareCommunication: Connection established
TelescopeTCP( "New Telescope 1" )::performReading:  bad packet size:  6144
Max thread count (Global Pool):  6
TelescopeTCP( "New Telescope 1" )::prepareCommunication: Attempting to connect to host "127.0.0.1" at port 10001
TelescopeTCP( "New Telescope 1" ): turning off Nagle algorithm.
TelescopeTCP( "New Telescope 1" )::prepareCommunication: Connection established
SlewDialog::loadPointsFromFile(): No points loaded. File is missing: "C:\\Users\\Lammert\\AppData\\Roaming\\Stellarium\\modules\\TelescopeControl\\points.json"

My client says :
Server is listening on 127.0.0.1:10001...
Connected to client
Sending data...
Sending data...
Sending data...
Send failed: 10053

My source code is :
`#define _WINSOCK_DEPRECATED_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <winsock2.h>
#include <windows.h>

#pragma comment(lib, "ws2_32.lib") // Link with ws2_32.lib

#define HOST "127.0.0.1"
#define PORT 10001

// Define the message structure
#pragma pack(push, 1) // Ensure no padding is added
typedef struct {
    int16_t length;   // 2 bytes
    int16_t type;     // 2 bytes
    int64_t time;     // 8 bytes
    uint32_t ra;      // 4 bytes, unsigned
    int32_t dec;      // 4 bytes, signed
    int32_t status;   // 4 bytes, signed
} Message;
#pragma pack(pop)

void initialize_message(Message* msg) {
    msg->length = htons(24);                     // Network byte order
    msg->type = htons(0);                        // Network byte order
    msg->time = _byteswap_uint64(0);             // Network byte order
    msg->ra = htonl(0x100000000 / 24);           // Network byte order
    msg->dec = htonl(0x40000000 / 90);           // Network byte order
    msg->status = htonl(0);                      // Network byte order
}

int main() {
    // Initialize Winsock
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        printf("WSAStartup failed: %d\n", WSAGetLastError());
        return 1;
    }

    // Create and initialize the message
    Message msg;
    initialize_message(&msg);

    // Create socket
    SOCKET server_socket, client_socket;
    struct sockaddr_in server_addr, client_addr;
    int client_len = sizeof(client_addr);

    server_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (server_socket == INVALID_SOCKET) {
        printf("Socket creation failed: %d\n", WSAGetLastError());
        WSACleanup();
        return 1;
    }

    // Bind to the specified host and port
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = inet_addr(HOST);
    server_addr.sin_port = htons(PORT);

    if (bind(server_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) == SOCKET_ERROR) {
        printf("Bind failed: %d\n", WSAGetLastError());
        closesocket(server_socket);
        WSACleanup();
        return 1;
    }

    // Listen for incoming connections
    if (listen(server_socket, 1) == SOCKET_ERROR) {
        printf("Listen failed: %d\n", WSAGetLastError());
        closesocket(server_socket);
        WSACleanup();
        return 1;
    }

    printf("Server is listening on %s:%d...\n", HOST, PORT);

    // Accept a connection
    client_socket = accept(server_socket, (struct sockaddr*)&client_addr, &client_len);
    if (client_socket == INVALID_SOCKET) {
        printf("Accept failed: %d\n", WSAGetLastError());
        closesocket(server_socket);
        WSACleanup();
        return 1;
    }

    printf("Connected to client\n");

    // Send data repeatedly with a delay
    while (1) {
        printf("Sending data...\n");
        if (send(client_socket, (const char*)&msg, sizeof(msg), 0) == SOCKET_ERROR) {
            printf("Send failed: %d\n", WSAGetLastError());
            break; // Exit the loop
        }
        Sleep(10000); // Delay of 10 second (10000 ms)
    }

    // Close the connections
    closesocket(client_socket);
    closesocket(server_socket);
    WSACleanup();

    return 0;
}`

Hope you guys can see the error.
gr. Lamko

**Comment by gzotti** [comment]
The purpose of the telescope control plugin is establishing outgoing connections to GOTO telescopes. Everything concerning connecting the TC plugin from external programs is IMHO out of scope. What do you want to achieve?

**Comment by lamko648** [comment]
I've got an dish that i want to control with goto commands and get the current postion of the dish in stellarium. I want it all running from a raspberrypi, the goto commands i can already receive on the pi.

**Reply by gzotti** [reply]
Then you need to implement a telescope client for the telescope plugin. (If its control unit does not already speak LX200/NexStar/INDI/ASCOM/RTS2.)

**Reply by lamko648** [reply]
Thats what I'm doing using this protocol : server->client:
MessageCurrentPosition (type = 0):  
LENGTH (2 bytes,integer): length of the message
TYPE   (2 bytes,integer)
TIME   (8 bytes,integer)
RA     (4 bytes,unsigned integer)
DEC    (4 bytes,signed integer)
STATUS (4 bytes,signed integer)

**Reply by lamko648** [reply]
So how can I help you better besides the logs ?

**Reply by gzotti** [reply]
I will be of no direct help here, sorry. You have the hardware and the motivation. It is just that the title of this question is misleading. You want to control your device with a control unit that runs on a RPi and need to implement one of the available protocols on it so that the TC plugin talks to your device. Nice. There are a few similar projects out there, and code here to study.

**Reply by lamko648** [reply]
I've found some other code which was doing the same as what i want it to be doing ? https://sourceforge.net/p/stellarium/discussion/278769/thread/97189be1/ on python but to no avail. Also been using very old versions of stellarium with qt5, nothing.
This is the message I'm sending : 00 18 00 00 00 00 00 00 00 00 00 00 10 00 00 00 00 cc cc cc 00 00 00 00
Should be correct, i'm very confused.
